### PR TITLE
Merge Refs and ScopedRefs

### PR DESCRIPTION
**Pull request by magnus-madsen**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@mlutze I pinged you to give you a chance to look at Typer

**Review by mlutze — COMMENTED** [review]
No tests? :)

A couple examples would help me understand better what's happening (and make sure I don't break it with the Kinder)

File: main/src/ca/uwaterloo/flix/language/phase/Typer.scala [hunk 1]
```diff
@@ -686,6 +686,17 @@ object Typer extends Phase[ResolvedAst.Root, TypedAst.Root] {
           resultEff = Type.mkAnd(eff1, eff2)
         } yield (constrs1 ++ constrs2, resultTyp, resultEff)
 
+      case ResolvedAst.Expression.LetRegion(sym, exp, evar, loc) =>
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Typer.scala:689]: This is more incidental than it is actually deep. Forget about `Region[b]` for now.

The real point is that inside Region there is a Boolean variable `l` called the lifetime of the region.

Then when a reference is allocated with `ref foo @ x` where `x` is a region with life time `l` we get a `ScopedRef[t, l]`.

Any operations on refs then use `l` as the effect (and not false).

(Note that refs allocated without a region have the "region" `false` which means all their operations are impure as before).

[the remark this answers is not in the document]
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Typer.scala:689]: Ok, makes sense.

**Review by mlutze — APPROVED** [review]

**Pull request merged**